apps/settings/js/modules/base/event_emitter.js

```diff
@@ -0,0 +1,121 @@
+/**
+ * EventEmitter implements basic event emitting logics. It also provides
+ * function that augments the ability to other objects. For example:
+ *  
+ *   function Module() {
+ *     this.EventEmitter.call(this, [eventName1, eventName2]);
+ *   }
+ *   EventEmitter.augment(Module.prototype);
+ *
+ * @module EventEmitter
+ */
+define(function() {
+  'use strict';
+
+  /**
+   * @class EventEmitter
+   * @returns {EventEmitter}
+   */
+  function EventEmitter(eventNames) {
+    if (eventNames && eventNames.length) {
+      this._eventListeners = {};
+      eventNames.forEach((eventName) => {
+        this._eventListeners[eventName] = [];
+      });
+    } else {
+      throw new Error('EventEmitter: no valid registered events');
+    }
+  }
+
+  /**
+   * Emit a event of a specific type and a given value.
+   *
+   * @access private
+   * @memberOf EventEmitter.prototype
+   * @param {String} eventName
+   * @param {Object} value
+   */
+  EventEmitter.prototype._emitEvent = function(eventName, value) {
+    var listeners = this._eventListeners[eventName];
+    if (listeners) {
+      var eventObj = {
+        type: eventName,
+        detail: value
+      };
+      listeners.forEach((listener) => {
+        listener.call(this, eventObj);
+      });
+    }
+  };
+
+  /**
+   * Add an event listener on an event. If listener is an object,
+   * object.handleEvent is called when the event emits.
+   *
+   * @access public
+   * @memberOf EventEmitter.prototype
+   * @param {String} eventName
+   * @param {(Function|Object)} listener
+   */
+  EventEmitter.prototype.addEventListener = function(eventName, listener) {
+    var listeners = this._eventListeners[eventName];
+    if (listener && listeners && listeners.indexOf(listener) < 0) {
+      var type = typeof listener;
+      switch(type) {
+        case 'function':
+          listeners.push(listener);
+          break;
+        case 'object':
```

Comment:
The `Object` case is for the case that we want to addEventListener on `this` object right ?